### PR TITLE
improve efficiency of the executor

### PR DESCRIPTION
**Pull request by glommer**
Improves efficiency of spawning tasks in the executor, by pre-computing the task queue index

Before:
```
cost to run task no task queue: 126ns
cost to run task in task queue: 300ns
```

After
```
cost to run task no task queue: 93ns
cost to run task in task queue: 257ns
```

No regressions found in any of the existing benchmarks. Semaphore and local channel show similar improvements for the contended case.

### CODE REVIEW

File: glommio/src/executor.rs [hunk 1]
```diff
@@ -781,15 +753,21 @@ impl LocalExecutor {
...
             .borrow()
             .active_executing
             .as_ref()
-            .map(|x| x.borrow().ex.clone())
-            .or_else(|| self.get_executor(&TaskQueueHandle { index: 0 }))
-            .unwrap();
-        Task(ex.spawn(future))
+            .map_or_else(
+                || self.get_queue(&TaskQueueHandle { index: 0 }),
+                |x| Some(x).cloned(),
```
Comment (zserik) [inline glommio/src/executor.rs:763]:
```suggestion
                |x| Some(x.clone()),
```

**Comment by glommer** [comment]
Thanks @zserik . I will leave this patch cooking for a bit longer. Due to a change that I consider too drastic (the fact that you can no longer spawn before run), I will bump the version to 0.3.0 once I do it.

File: glommio/src/executor.rs (hunk 1 above)
Comment (zserik) [reply glommio/src/executor.rs]: Ah, I didn't realise that `x` is a `Rc`... so it should be `|x| Some((*x).clone())` instead. Sorry.

File: glommio/src/executor.rs [hunk 2]
```diff
@@ -781,15 +753,21 @@ impl LocalExecutor {
...
             .borrow()
             .active_executing
             .as_ref()
-            .map(|x| x.borrow().ex.clone())
-            .or_else(|| self.get_executor(&TaskQueueHandle { index: 0 }))
-            .unwrap();
-        Task(ex.spawn(future))
+            .map_or_else(
+                || self.get_queue(&TaskQueueHandle { index: 0 }),
+                |&x| Some(x.clone()),
```
Comment (zserik) [inline glommio/src/executor.rs:763]: Why do we even need a `clone()` here?

**Comment by zserik** [comment]
I created a PR to `glommer:executor` which contains a better suggestion, as I've made an error in the last one...

File: glommio/src/executor.rs [hunk 3]
```diff
@@ -915,6 +896,24 @@ impl LocalExecutor {
         }
     }
 
+    // this happens if we spawn() into the executor before we run().
+    // I am considering making this illegal, but in the meantime...
+    fn activate_pre_spawned_queues(&self) {
```
Comment (zserik) [inline glommio/src/executor.rs:901]: If we can no longer `spawn` before `run`, why is this `fn` here?

Comment (glommer) [reply glommio/src/executor.rs:901]: should have been deleted. It was added to handle the case where we spawn before run and I forgot to delete. Thanks!

File: glommio/Cargo.toml [hunk 4]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "glommio"
-version = "0.2.0"
+version = "0.3.0"
```
Comment (zserik) [inline glommio/Cargo.toml:3]: `examples/Cargo.toml` `[dependencies] glommio = ...` should be updated to match the new version number.

Comment (glommer) [reply glommio/Cargo.toml:3]: you are faster than the CI =p

File: glommio/src/executor.rs (hunk 3 above)
Comment (glommer) [reply glommio/src/executor.rs:896]: I am now wondering if I haven't simply uploaded the wrong version entirely? The spawn functions are still public =(

**Comment by glommer** [comment]
ETA to publish a new crate is ~ 2 weeks, hopefully when the network API stabilizes

**Pull request merged**